src/main/java/com/conveyal/r5/analyst/Grid.java

```diff
@@ -126,96 +126,154 @@ public Grid (int zoom, int width, int height, int north, int west) {
+        int x;
+        int y;
+        double weight;
+
+        private PixelWeight (int x, int y, double weight){
+            this.x = x;
+            this.y = y;
+            this.weight = weight;
+        }
+
+        public int getX() {
+            return x;
+        }
+        public int getY() {
+            return y;
+        }
+        public double getWeight() {
+            return weight;
+        }
+    }
+
+    /**
+     * A functional interface (callback interface) that processes one pixel weight at a time.
+     * This avoids storing massive lists of intermediate results that have been known to eat memory on the server.
+     * This happens when processing geometries that contain a huge amount of pixels, which a user might include in
+     * a shapefile.
+     */
+    public interface PixelWeightCallback {
+        void handlePixelWeight (int x, int y, double weight);
+    }
+
+    /**
+     * Stream pixel weights using a callback, rather than keeping them in memory.
+     */
+    public void streamPixelWeights (Geometry geometry, boolean relativeToPixels, PixelWeightCallback pixelWeightCallback) {
+        getPixelWeights(geometry, relativeToPixels, pixelWeightCallback);
+    }
+
     /**
      * Version of getPixelWeights which returns the weights as relative to the total area of the input geometry (i.e.
      * the weight at a pixel is the proportion of the input geometry that falls within that pixel.
      */
-    public TObjectDoubleMap<int[]> getPixelWeights (Geometry geometry) {
-        return getPixelWeights(geometry, false);
+    public ArrayList<PixelWeight> getPixelWeights (Geometry geometry) {
+        return getPixelWeights(geometry, false, null);
     }
 
     /**
-     * Get the proportions of an input polygon feature that overlap each grid cell, in the format [x, y] => weight.
-     * This weight object can then be fed into the incrementFromPixelWeights function to actually burn a polygon into the
+     * Get the proportions of an input polygon feature that overlap each grid cell, for use in List(x, y, weight)
+     * These weight lists can then be fed into the incrementFromPixelWeights function to actually burn a polygon into the
      * grid.
      *
      * If relativeToPixels is true, the weights are the proportion of the pixel that is covered. Otherwise they are the
-     * portion of this polygon which is within the given grid cell. If using incrementPixelWeights, this should be set to
+     * portion of this polygon which is within the given pixel. If using incrementPixelWeights, this should be set to
      * false.
      *
-     * This returns a map from int arrays containing the coordinates to the weight. I believe int arrays in hashmaps are
-     * hashed based on memory location, meaning lookups will fail. However, we only use this in a streaming fashion, looping
-     * over all values, which is harmless.
+     * This used to return a map from int arrays containing the coordinates to the weight.
```

Comment:
It's pretty unconventional to have a function that returns an empty collection when it has been given a callback, but otherwise puts things in the collection. It would seem a little safer to me if the function returned null when it used the callback. It took me a moment to realize this is an internal private method that's called by two different public methods with different signatures, which is a good way to do it, but deserves a very clear explanation in the Javadoc.